### ISSUE TITLE
setGlobalOptions must be called before usages of @modelOptions()

### ISSUE THREAD

**Issue by richard-jfc** [typegoose/typegoose#939]
## Versions

- System: linux
- NodeJS: 18.19.0
- Typescript: 5.3.3

- Compiler / Transpiler: ts-jest 
- Typegoose(NPM): 12.4.0
- mongoose: 8.3.4
- mongodb: 6.0.14

## What is the Problem?

Classes that use `@modelOptions()` are not respecting `setGlobalOptions`. This is particular unobvious when the class only uses `@modelOptions` because it extends `TimeStamps`.

## Code Example

```ts
import {
  buildSchema,
  prop,
  setGlobalOptions,
  Severity,
} from '@typegoose/typegoose';
import { TimeStamps } from '@typegoose/typegoose/lib/defaultClasses';

setGlobalOptions({ options: { allowMixed: Severity.ERROR } });

class User extends TimeStamps {
  @prop()
  username: string | undefined; // This will result in a Mixed type
}

buildSchema(User); // Only logs a warning, should throw an error
```

## Do you know *why* it happens?

`@modelOptions()` attempts to apply the global options immediately, it doesn't wait for `buildSchema` to be called. Importing `TimeStamps` calls `@modelOptions()`, which happens before the `setGlobalOptions` call. 

## Expected Behavior
global options should be applied when `buildSchema` is called, not when `@modelOptions` is called.

---

## Work around 
Relies on imports loading in a specific order
```ts
// globalTypegooseOptions.ts
import { setGlobalOptions, Severity } from '@typegoose/typegoose';

setGlobalOptions({ options: { allowMixed: Severity.ERROR } });

//---
// main.ts
import './globalTypegooseOptions';
import { buildSchema, prop } from '@typegoose/typegoose';
import { TimeStamps } from '@typegoose/typegoose/lib/defaultClasses';

class User extends TimeStamps {
  @prop()
  username: string | undefined; // This will result in a Mixed type
}

buildSchema(User); // Correctly throws an error

```

**Comment by hasezoey** [comment]
yes, `@modelOptions` currently directly merges on decorator call (and on first call applies global options).

the only fix to this i can image is re-working how modelOptions works by storing all arguments in a array and on `buildSchema` apply / merge them in-order, which i would consider it being a breaking change to how it currently works.

**Comment by lantw44** [comment]
We are still using Typegoose 11.7.1 and affected by the bug. The import order workaround doesn't work because importing `@typegoose/typegoose` causes `modelOptions` on `defaultClasses.TimeStamps` to be executed. Even the `TG_ALLOW_MIXED` environment variable can't work because `parseENV` is called after the `defaultClasses` export. Therefore, we have to import internal files directly in order to work around the issue:
```typescript
import { setGlobalOptions } from '@typegoose/typegoose/lib/globalOptions.js';
import { Severity } from '@typegoose/typegoose/lib/internal/constants.js';

setGlobalOptions({ options: { allowMixed: Severity.ERROR } });
```

**Comment by hasezoey** [comment]
> Even the TG_ALLOW_MIXED environment variable can't work because parseENV is called after the defaultClasses export

At least this issue is now fixed with [`12.9.1`](https://github.com/typegoose/typegoose/releases/tag/v12.9.1) and [`11.8.1`](https://github.com/typegoose/typegoose/releases/tag/v11.8.1), thanks for reporting.

**Comment by lantw44** [comment]
> At least this issue is now fixed with [`12.9.1`](https://github.com/typegoose/typegoose/releases/tag/v12.9.1) and [`11.8.1`](https://github.com/typegoose/typegoose/releases/tag/v11.8.1), thanks for reporting.

Thank you. Now `TG_ALLOW_MIXED` works.
